### PR TITLE
feat: add casdoor as saml idp support 

### PR DESCRIPTION
**Pull request by Steve0x2a**
Now it is necessary to perform sp connection test, so it is set to draft state
Signed-off-by: 0x2a <stevesough@gmail.com>

Fix: https://github.com/casdoor/casdoor/issues/405

### CODE REVIEW

**Comment by hsluoyz** [comment]
@seriouszyx plz review

**Comment by qwqcode** [comment]
nice work!

**Comment by hsluoyz** [comment]
@seriouszyx plz review

File: web/src/auth/LoginPage.js [hunk 1]
```diff
@@ -157,30 +172,39 @@ class LoginPage extends React.Component {
...
+      }
+  
       AuthBackend.login(values, oAuthParams)
         .then((res) => {
           if (res.status === 'ok') {
             const responseType = values["type"];
             if (responseType === "login") {
               Util.showMessage("success", `Logged in successfully`);
-
+
```
Comment (hsluoyz) [inline web/src/auth/LoginPage.js:191]: no change

File: web/src/auth/LoginPage.js [hunk 2]
```diff
@@ -157,30 +172,39 @@ class LoginPage extends React.Component {
...
               Util.showMessage("success", `Logged in successfully`);
-
+  
               const link = Setting.getFromLink();
               Setting.goToLink(link);
             } else if (responseType === "code") {
               const code = res.data;
               const concatChar = oAuthParams?.redirectUri?.includes('?') ? '&' : '?';
-
+
```
Comment (hsluoyz) [inline web/src/auth/LoginPage.js:197]: no change

File: web/src/auth/LoginPage.js [hunk 3]
```diff
@@ -157,30 +172,39 @@ class LoginPage extends React.Component {
...
+  
               if (Setting.hasPromptPage(application)) {
                 AuthBackend.getAccount("")
                   .then((res) => {
                     let account = null;
                     if (res.status === "ok") {
                       account = res.data;
                       account.organization = res.data2;
-
+
```
Comment (hsluoyz) [inline web/src/auth/LoginPage.js:205]: no change

File: web/src/auth/LoginPage.js [hunk 4]
```diff
@@ -157,30 +172,39 @@ class LoginPage extends React.Component {
...
                   .then((res) => {
                     let account = null;
                     if (res.status === "ok") {
                       account = res.data;
                       account.organization = res.data2;
-
+  
                       this.onUpdateAccount(account);
-
+
```
Comment (hsluoyz) [inline web/src/auth/LoginPage.js:207]: no change

File: web/src/auth/LoginPage.js [hunk 5]
```diff
@@ -193,11 +217,15 @@ class LoginPage extends React.Component {
               } else {
                 Setting.goToLink(`${oAuthParams.redirectUri}${concatChar}code=${code}&state=${oAuthParams.state}`);
               }
-
+
```
Comment (hsluoyz) [inline web/src/auth/LoginPage.js:220]: no change

File: web/src/auth/Util.js [hunk 6]
```diff
@@ -101,8 +101,9 @@ export function getOAuthGetParameters(params) {
   const nonce = getRefinedValue(queries.get("nonce"))
```
Comment (hsluoyz) [inline web/src/auth/Util.js:101]: ; for all lines

File: web/src/auth/AuthCallback.js [hunk 7]
```diff
@@ -49,6 +49,10 @@ class AuthCallback extends React.Component {
       const realRedirectUri = innerParams.get("redirect_uri");
       // Casdoor's own login page, so "code" is not necessary
       if (realRedirectUri === null) {
+        const samlRequest = innerParams.get("SAMLRequest");
```
Comment (hsluoyz) [inline web/src/auth/AuthCallback.js:52]: SAML -> saml

File: web/src/auth/AuthCallback.js [hunk 8]
```diff
@@ -92,6 +96,7 @@ class AuthCallback extends React.Component {
     const applicationName = innerParams.get("application");
     const providerName = innerParams.get("provider");
     const method = innerParams.get("method");
+    const samlRequest = innerParams.get("SAMLRequest");
```
Comment (hsluoyz) [inline web/src/auth/AuthCallback.js:99]: SAML -> saml

File: web/src/auth/Util.js [hunk 9]
```diff
@@ -101,8 +101,9 @@ export function getOAuthGetParameters(params) {
...
   const challengeMethod = getRefinedValue(queries.get("code_challenge_method"))
   const codeChallenge = getRefinedValue(queries.get("code_challenge"))
-  
-  if (clientId === undefined || clientId === null || clientId === "") {
+  const samlRequest = getRefinedValue(queries.get("SAMLRequest"))
```
Comment (hsluoyz) [inline web/src/auth/Util.js:104]: SAML -> saml

File: object/samlidp.go [hunk 10]
```diff
@@ -0,0 +1,274 @@
+// Copyright 2022 The Casdoor Authors. All Rights Reserved.
```
Comment (hsluoyz) [inline object/samlidp.go:1]: this file -> saml_idp.go

saml.go -> saml_sp.go

File: controllers/saml.go [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+	host := c.Ctx.Request.Host
+	paramApp := c.Input().Get("application")
+	application := object.GetApplication(paramApp)
+	if application == nil {
+		c.ResponseError(fmt.Sprintf("err: application %s not found", paramApp))
+	}
+	metadata, _ := object.GetSamlMeta(application, host)
+	c.Data["xml"] = metadata
+	c.ServeXML()
+}
```
Comment (hsluoyz) [inline controllers/saml.go:33]: new line

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Review by hsluoyz — APPROVED** [review]

**Pull request merged**